### ISSUE TITLE
The validation loss is rising and fluctuating, is that a regular situation?

### ISSUE THREAD

**Issue by Buckingham0805** [auspicious3000/SpeechSplit#11]
Greetings, thanks for such a good project.
In my experiment, i used the same dataset VCTK as yours, and  i had only trained for 68000-steps. The log of my experiment like this:

![311600500287_ pic](https://user-images.githubusercontent.com/26041589/93661685-42b8e380-fa8c-11ea-914c-273da045dbc6.jpg)

I noticed that the validation loss is rising and the training loss also has some fluctuating peaks. Is it a normal phenomenon？

thank you in advance :)

**Comment by auspicious3000** [comment]
The provided training data is very small for code verification purposes only.

**Comment by Buckingham0805** [comment]
> The provided training data is very small for code verification purposes only.

Indeed. But the image above comes from experiment which using my own VCTK data. There were 20 speakers in my VCTK corpus, 80% of utterances were used in training steps and 10% for validation. I orgnized the data structure in the same form of provided .pkl file.
Is there sth wrong happened in my experiment?

**Comment by auspicious3000** [comment]
There might be something wrong with your validation data. The validation loss should be around 30.

**Comment by Buckingham0805** [comment]
> There might be something wrong with your validation data. The validation loss should be around 30.

Thanks for your answer. I will check my preprocessing code. By the way, is it a right way for me to generate validation data in the **same structure** as that in your demo.pkl? 
Like this:
![image](https://user-images.githubusercontent.com/26041589/93725993-013c4b80-fbe6-11ea-9fba-535448f0a11e.png)

[Speaker_Name , One-hot , [Mel, normed-F0, length, utterance_name] ]

This is how i did now.

Thanks again for your answer!

**Comment by auspicious3000** [comment]
The format is correct.

**Comment by inconnu11** [comment]
Did you normalize the Mel spectrogram? What's the range of the Mel spec?

**Comment by niu0717** [comment]
The code require audio sample rate is 16kHz, but the VCTK is 48kHz. 
How can i run make_spect_f0.py with the origin VCTK corpus ?

**Comment by niu0717** [comment]
Hi,  i am grateful if you can tell me how to restructure the demo.pkl.  thx :) 
https://github.com/auspicious3000/SpeechSplit/blob/10ed8b9e25cce6c9a077e27ca175ba696b7df597/solver.py#L16

**Comment by c1a1o1** [comment]
Can you make the right demo.pkl file?

**Comment by c1a1o1** [comment]
@auspicious3000  Can you make the right demo.pkl file?

**Comment by auspicious3000** [comment]
@c1a1o1 Please clearly state your question. Please do NOT flood other issues.

**Comment by jamesliu** [comment]
Thanks for good paper and project. My experiment is similar to Buckingham. My validation loss fluctuates around 100 after 70K iterations. I haven't figured out what is wrong with my experiment. It would be great in case any suggestions.
![Screen Shot 2020-11-07 at 10 19 29 PM](https://user-images.githubusercontent.com/41763/98458650-da12ec80-2147-11eb-8095-2d040c97c3a5.png)

**Comment by auspicious3000** [comment]
@jamesliu This looks like over-fitting to me. Make sure you use a large training set and the validation speakers are in the training set.

**Comment by jamesliu** [comment]
@auspicious3000  Yes. Thank you for pointing this out. After starting to use full P226 and P231 in VCTK corpus. The training and validation charts look reasonable now, but the reconstruction from the trained model is not good for the demo data. Do I need to run more iterations or change to other optimizer in stead of Adam? Thanks.

G training/validation charts
--------
![Screen Shot 2020-11-11 at 7 46 40 PM](https://user-images.githubusercontent.com/41763/98893272-f750ff00-2456-11eb-934f-03687b7fa74f.png)
 
reconstruction from model
--------------------
![Screen Shot 2020-11-11 at 7 39 07 PM](https://user-images.githubusercontent.com/41763/98893064-87427900-2456-11eb-81b0-ba1b803712de.png)

**Comment by auspicious3000** [comment]
@jamesliu Your training set is actually very small, which has only 30 mins of data.  Also, the "demo data" needs to be consistent with the training data.

**Comment by jamesliu** [comment]
@auspicious3000 Thanks for your suggestion, I have trained the 80 speakers(P225~P304) in VCTK dataset(due to onehot size is 80) on 2080Ti GPU for 2 days, the result becomes better, but not good enough. Is my training set big enough to avoid overfitting? How many days or iterations should I expect to get a good result? 
Please ignore the sudden rise of errors in validation set, I took more examples for validation after 170K.
![Screen Shot 2020-11-14 at 7 38 12 AM](https://user-images.githubusercontent.com/41763/99150938-8e32dc80-264c-11eb-8454-b562a5a1f734.png)

hparams
======
# Default hyperparameters:
hparams = HParams(
    # synthesis
    builder = 'wavenet',
    
    # model   
    freq = 8,
    dim_neck = 8,
    freq_2 = 8,
    dim_neck_2 = 1,
    freq_3 = 8,
    dim_neck_3 = 32,
    
    dim_enc = 512,
    dim_enc_2 = 128,
    dim_enc_3 = 256,
    
    dim_freq = 80,
    dim_spk_emb = 82,
    dim_f0 = 257,
    dim_dec = 512,
    len_raw = 128,
    chs_grp = 16,
    
    # interp
    min_len_seg = 19,
    max_len_seg = 32,
    min_len_seq = 64,
    max_len_seq = 128,
    max_len_pad = 192,
    
    # data loader
    root_dir = '/data/music/speech_split/assets/spmel',
    feat_dir = '/data/music/speech_split/assets/raptf0',
    batch_size = 128,
    mode = 'train',
    shuffle = True,
    num_workers = 10,
    samplier = 8,
    #optimizer = 'RangerLars'
    optimizer = 'Adam'
)

**Comment by auspicious3000** [comment]
There are many contributing factors to output quality.  It is hard to tell from the information you provided. @jamesliu

**Comment by FurkanGozukara** [comment]
Could you guys check my problem?

I am trying to achieve very simple train and apply model

@jamesliu @Buckingham0805 @niu0717 @inconnu11 

https://github.com/auspicious3000/SpeechSplit/issues/28

**Comment by tejuafonja** [comment]
Hi James, 

I followed your updates and wonder if you continued with the experiments. I plan to run some experiments with another dataset and would like to learn from your experiments thus far. 

Did the results become better?
Did you increase the training size?
Did you increase the training time?
Did you change the hyperparameter? 
Did you try out other techniques?

Thanks and hope to hear from you!

**Comment by 3139725181** [comment]
> validation

Hi, I noticed that the length of most data in VCTK is too long relative to the required input. Have you processed the validation set, like MyCollator in data_loader?

**Comment by 3139725181** [comment]
> @auspicious3000 Thanks for your suggestion, I have trained the 80 speakers(P225~P304) in VCTK dataset(due to onehot size is 80) on 2080Ti GPU for 2 days, the result becomes better, but not good enough. Is my training set big enough to avoid overfitting? How many days or iterations should I expect to get a reasonable result? Any suggestion for hyper parameters? Thanks.
> Please ignore the sudden rise of errors in validation set, I took more examples for validation after 170K.
> ![Screen Shot 2020-11-14 at 7 38 12 AM](https://user-images.githubusercontent.com/41763/99150938-8e32dc80-264c-11eb-8454-b562a5a1f734.png)
> 
> # hparams
> # Default hyperparameters:
> hparams = HParams(
> # synthesis
> builder = 'wavenet',
> 
> ```
> # model   
> freq = 8,
> dim_neck = 8,
> freq_2 = 8,
> dim_neck_2 = 1,
> freq_3 = 8,
> dim_neck_3 = 32,
> 
> dim_enc = 512,
> dim_enc_2 = 128,
> dim_enc_3 = 256,
> 
> dim_freq = 80,
> dim_spk_emb = 82,
> dim_f0 = 257,
> dim_dec = 512,
> len_raw = 128,
> chs_grp = 16,
> 
> # interp
> min_len_seg = 19,
> max_len_seg = 32,
> min_len_seq = 64,
> max_len_seq = 128,
> max_len_pad = 192,
> 
> # data loader
> root_dir = '/data/music/speech_split/assets/spmel',
> feat_dir = '/data/music/speech_split/assets/raptf0',
> batch_size = 128,
> mode = 'train',
> shuffle = True,
> num_workers = 10,
> samplier = 8,
> #optimizer = 'RangerLars'
> optimizer = 'Adam'
> ```
> 
> )

Hi, I noticed that the length of most data in VCTK is too long relative to the required input. Have you processed the validation set, like MyCollator in data_loader?

**Comment by anon-squid** [comment]
I have the same question as @3139725181. How can we use longer audio files in the validation set?

**Comment by jixinya** [comment]
> @auspicious3000 Thanks for your suggestion, I have trained the 80 speakers(P225~P304) in VCTK dataset(due to onehot size is 80) on 2080Ti GPU for 2 days, the result becomes better, but not good enough. Is my training set big enough to avoid overfitting? How many days or iterations should I expect to get a reasonable result? Any suggestion for hyper parameters? Thanks. Please ignore the sudden rise of errors in validation set, I took more examples for validation after 170K. ![Screen Shot 2020-11-14 at 7 38 12 AM](https://user-images.githubusercontent.com/41763/99150938-8e32dc80-264c-11eb-8454-b562a5a1f734.png)
> 
> # hparams
> # Default hyperparameters:
> hparams = HParams( # synthesis builder = 'wavenet',
> 
> ```
> # model   
> freq = 8,
> dim_neck = 8,
> freq_2 = 8,
> dim_neck_2 = 1,
> freq_3 = 8,
> dim_neck_3 = 32,
> 
> dim_enc = 512,
> dim_enc_2 = 128,
> dim_enc_3 = 256,
> 
> dim_freq = 80,
> dim_spk_emb = 82,
> dim_f0 = 257,
> dim_dec = 512,
> len_raw = 128,
> chs_grp = 16,
> 
> # interp
> min_len_seg = 19,
> max_len_seg = 32,
> min_len_seq = 64,
> max_len_seq = 128,
> max_len_pad = 192,
> 
> # data loader
> root_dir = '/data/music/speech_split/assets/spmel',
> feat_dir = '/data/music/speech_split/assets/raptf0',
> batch_size = 128,
> mode = 'train',
> shuffle = True,
> num_workers = 10,
> samplier = 8,
> #optimizer = 'RangerLars'
> optimizer = 'Adam'
> ```
> 
> )

Hi, I use the whole VCTK  datset to trian, but the validation loss fluctuates around 70 and then rise. I wonder how you generate the validation data, the same as the training data?

**Comment by auspicious3000** [comment]
@jixinya Yes, the validation data is just a separate partition of the training data.

**Comment by 9527950** [comment]
> @auspicious3000 Thanks for your suggestion, I have trained the 80 speakers(P225~P304) in VCTK dataset(due to onehot size is 80) on 2080Ti GPU for 2 days, the result becomes better, but not good enough. Is my training set big enough to avoid overfitting? How many days or iterations should I expect to get a reasonable result? Any suggestion for hyper parameters? Thanks. Please ignore the sudden rise of errors in validation set, I took more examples for validation after 170K. ![Screen Shot 2020-11-14 at 7 38 12 AM](https://user-images.githubusercontent.com/41763/99150938-8e32dc80-264c-11eb-8454-b562a5a1f734.png)
> 
> # hparams
> # Default hyperparameters:
> hparams = HParams( # synthesis builder = 'wavenet',
> 
> ```
> # model   
> freq = 8,
> dim_neck = 8,
> freq_2 = 8,
> dim_neck_2 = 1,
> freq_3 = 8,
> dim_neck_3 = 32,
> 
> dim_enc = 512,
> dim_enc_2 = 128,
> dim_enc_3 = 256,
> 
> dim_freq = 80,
> dim_spk_emb = 82,
> dim_f0 = 257,
> dim_dec = 512,
> len_raw = 128,
> chs_grp = 16,
> 
> # interp
> min_len_seg = 19,
> max_len_seg = 32,
> min_len_seq = 64,
> max_len_seq = 128,
> max_len_pad = 192,
> 
> # data loader
> root_dir = '/data/music/speech_split/assets/spmel',
> feat_dir = '/data/music/speech_split/assets/raptf0',
> batch_size = 128,
> mode = 'train',
> shuffle = True,
> num_workers = 10,
> samplier = 8,
> #optimizer = 'RangerLars'
> optimizer = 'Adam'
> ```
> 
> )

I also used all the voices from p225-p246, but my validation set loss function was oscillating upwards. And I use the trained model for conversion and it works poorly, and strangely the content of the sentences has changed, do you know what is causing this?
